packages/api/src/controllers/animalBatchController.js

```diff
@@ -0,0 +1,150 @@
+  getFarmAnimalBatches() {
+    return async (req, res) => {
+      try {
+        const { farm_id } = req.headers;
+        const rows = await AnimalBatchModel.query()
+          .where({ farm_id })
+          .whereNotDeleted()
+          .withGraphFetched('animal_batch_sex_detail');
+        return res.status(200).send(rows);
+      } catch (error) {
+        console.error(error);
+        return res.status(500).json({
+          error,
+        });
+      }
+    };
+  },
+
+  addAnimalBatches() {
+    return async (req, res) => {
+      const trx = await transaction.start(Model.knex());
+
+      try {
+        const { farm_id } = req.headers;
+        const result = [];
+
+        if (!Array.isArray(req.body)) {
+          await trx.rollback();
+          return res.status(400).send('Request body should be an array');
+        }
+
+        for (const animalBatch of req.body) {
+          if (animalBatch.custom_type_id) {
+            const customType = await CustomAnimalTypeModel.query().findById(
+              animalBatch.custom_type_id,
+            );
+            if (customType && customType.farm_id !== farm_id) {
+              await trx.rollback();
+              return res.status(403).send('Forbidden custom type does not belong to this farm');
+            }
+          }
+
+          if (animalBatch.default_breed_id && animalBatch.default_type_id) {
+            const defaultBreed = await DefaultAnimalBreedModel.query().findById(
+              animalBatch.default_breed_id,
+            );
+
+            if (defaultBreed && defaultBreed.default_type_id !== animalBatch.default_type_id) {
+              await trx.rollback();
+              return res.status(400).send('Breed does not match type');
+            }
+          }
+
+          if (animalBatch.custom_breed_id) {
+            const customBreed = await CustomAnimalBreedModel.query()
+              .whereNotDeleted()
+              .findById(animalBatch.custom_breed_id);
+
+            if (customBreed && customBreed.farm_id !== farm_id) {
+              await trx.rollback();
+              return res.status(400).send('Forbidden custom breed does not belong to this farm');
+            }
+
+            if (
+              customBreed.default_type_id &&
+              customBreed.default_type_id !== animalBatch.default_type_id
+            ) {
+              await trx.rollback();
+              return res.status(400).send('Breed does not match type');
+            }
+
+            if (
+              customBreed.custom_type_id &&
+              customBreed.custom_type_id !== animalBatch.custom_type_id
+            ) {
+              await trx.rollback();
+              return res.status(400).send('Breed does not match type');
+            }
+          }
+
+          if (animalBatch.count < 2) {
+            await trx.rollback();
+            return res.status(400).send('Batch count must be greater than 1');
+          }
+
+          if (animalBatch.animal_batch_sex_detail) {
```

Comment:
NIT but I think we should call this `sex_detail` in the request since the `animal_batch` is implicit (as it is in the other fields), although we can rename it here to have the right shape for the graph insert

Reply:
yeah.. it is long. The relation mapping name can technically be named anything so we could do that. I just went with our usual convention. If I change this I would prefer to change it all the way up to the model.

Reply:
Yup I'd change it all the way 🙂

Reply:
Cool done!